pylint/config.py

```diff
@@ -594,11 +594,69 @@ def parse(self, to_parse, config):
                     type_ = definition.get('type')
                     validator = VALIDATORS.get(type_, lambda x: x)
                     value = validator(value)
-                config.set_option(option, value)
+                configuration.set_option(option, value)
+
+    def parse(self, to_parse, config):
+        self._parser.read(to_parse)
+        self.apply_to_configuration(config)
 
     def write(self, stream=sys.stdout):
-        # TODO: Check if option descriptions are written out
-        self._parser.write(stream)
+        """
+        Write out config to stream.
+
+        Includes option help and options with default values as
+        comments.
+        """
+        # We can't reuse self._parser because it doesn't have the help
+        # comments. allow_no_value lets us add comments as keys.
+        write_parser = configparser.ConfigParser(
+            default_section='MASTER',
+            allow_no_value=True,
+        )
+
+        # This makes keys case sensitive, needed for preserving comment
+        # formatting.
+        write_parser.optionxform = str
+
+        configuration = Configuration()
+        self.apply_to_configuration(configuration)
```

Comment:
apply_to_configuration uses `self._parser` (as did the original code). This might mean we can only write config that came from a config file (or got into `self._parser` some other way I guess), not another source. Is this the intended functionality?

Reply:
I think we're fine to preserve existing behaviour here.